tree/ntuple/v7/inc/ROOT/RColumnElement.hxx

```diff
@@ -683,6 +698,74 @@ public:
    }
 };
 
+namespace BitPacking {
+
+using Word_t = std::uintmax_t;
+constexpr std::size_t kBitsPerWord = sizeof(Word_t) * 8;
+
+/// Returns the minimum safe size (in bytes) of a buffer that is intended to be used as a destination for PackBits
+/// or a source for UnpackBits.
+/// Passing a buffer that's less than this size will cause invalid memory reads and writes.
+constexpr std::size_t MinBufSize(std::size_t count, std::size_t nDstBits)
+{
+   return (count * nDstBits + 7) / 8;
+}
+
+/// Tightly packs `count` items of size `sizeofSrc` contained in `src` into `dst` using `nDstBits` per item.
+/// It must be  `0 < sizeofSrc <= 8`  and  `0 < nDstBits <= sizeofSrc * 8`.
+/// The extra least significant bits are dropped (assuming LE ordering of the items in `src`).
+/// Note that this function doesn't do any byte reordering for you.
+/// IMPORTANT: the size of `dst` must be at least `MinBufSize(count, nBitBits)`
+void PackBits(void *dst, const void *src, std::size_t count, std::size_t sizeofSrc, std::size_t nDstBits);
+
+/// Undoes the effect of `PackBits`. The bits that were truncated in the packed representation
+/// are filled with zeroes.
+/// `src` and `dst` must be at least `MinBufSize(count, nDstBits)` bytes long.
+void UnpackBits(void *dst, const void *src, std::size_t count, std::size_t sizeofDst, std::size_t nSrcBits);
+
+}; // namespace BitPacking
+
+template <>
+class RColumnElement<float, EColumnType::kReal32Trunc> : public RColumnElementBase {
+public:
+   static constexpr bool kIsMappable = false;
+   static constexpr std::size_t kSize = sizeof(float);
+
+   RColumnElement() : RColumnElementBase(kSize, kReal32TruncBitsMax) {}
+
+   void SetBitsOnStorage(std::size_t bitsOnStorage) final
+   {
+      R__ASSERT(bitsOnStorage >= kReal32TruncBitsMin && bitsOnStorage <= kReal32TruncBitsMax);
+      fBitsOnStorage = bitsOnStorage;
+   }
+
+   bool IsMappable() const final { return kIsMappable; }
+
+   void Pack(void *dst, void *src, std::size_t count) const final
+   {
+      R__ASSERT(GetPackedSize(count) == BitPacking::MinBufSize(count, fBitsOnStorage));
+
+#if R__LITTLE_ENDIAN == 0
+      auto bswapped = std::make_unique<float[]>(count);
```

Comment:
I originally had the byte swap inside the Un/PackBits functions, then I extracted it because I wasn't sure if we would only want to use those functions with 32 bit values.
I guess for now it's the case, so it's probably worth doing the change since it avoids an allocation.